src/ceph-volume/ceph_volume/util/device.py

```diff
@@ -264,21 +262,15 @@ def used_by_ceph(self):
                    if lv.tags.get("ceph.type") in ["data", "block"]]
         return any(osd_ids)
 
-
-    @property
-    def is_valid(self):
-        def reject_device(item, value, reason):
-            try:
-                if self.sys_api[item] == value:
-                    self._rejected_reasons.append(reason)
-            except KeyError:
-                pass
-        reject_device('removable', 1, 'removable')
-        reject_device('ro', 1, 'read-only')
-        reject_device('locked', 1, 'locked')
-
-        self._valid = len(self._rejected_reasons) == 0
-        return self._valid
+    def _is_available(self):
+        reasons = [
+            ('removable', 1, 'removable'),
+            ('ro', 1, 'read-only'),
+            ('locked', 1, 'locked'),
+        ]
+        rejected = [reason for (k, v, reason) in reasons if
+                    self.sys_api.get(k, '') == v]
+        return len(rejected) == 0, rejected
```

Comment:
this looks a bit odd, how is `len(rejected)` going to ever be equal to a tuple?

Reply:
oh, never mind, I get it now, you are actually returning a tuple

Reply:
I don't think there is a need to have this return a tuple. If it returned a list with the rejected reasons only it would meet the same goal. The method name would need to change (`_rejected_reasons()` maybe?) and then other methods can set the properties depending on the content.

`self.available` then would be `bool(len(self._rejected_reasons()))` and `self.rejected_reasons` would just be the return value. That implementation would be similar to what @ErwanAliasr1 did initially.

Reply:
The advantage of returning a tuple (and assigning the components) is that its atomic. The ```rejected``` list must be created before ```available``` makes sense. Returning a tuple and assigning it on object creation leaves no potential for error (say through future changes or something).
Are there any downsides to doing it this way?

Reply:
The downside for me is that it is an unexpected return given the name of the method. 

"is available" reads to me like a boolean, but this doesn't tell you if it is available, it gives you the reasons aside from the boolean.

The previous implementation kept that contract well.

When you say 'atomic' you mean that one method does it all for you I think? I would rather see single-purpose methods (and functions) when possible, specially when the name is implying it

Reply:
I'm more than happy to change the name.

As for the tuple assignment: imho this simply eliminates the chance of ever introducing an error by reordering assignments, i.e. assign the available property before rejected_reasons is populated.